### ISSUE TITLE
More than one attach target in target groups

### ISSUE THREAD

**Issue by stivi21** [terraform-aws-modules/terraform-aws-alb#316]
## Is your request related to a problem? Please describe.
In module version 8.x was option **targets** in **target_groups** configuration, for attach destination target or targets. In new 9.x module, as I understand, I can configure only 1 target per target group.

## Describe the solution you'd like.
I need to attach few targets (servers) in 1 target group (fro HA configuration), so I would appreciate if you will add option for attach more than 1 destination in target group.

**Comment by bryantbiggs** [comment]
The recommendation there for HA would be to attach an autoscaling group

**Comment by pat-s** [comment]
@bryantbiggs  Could you give an example how this could be done? I've tried using an ASG ID and arn for `target_id` but both were rejected. Example shows using ARN:

```tf
  target_groups = {
    instance_http = {
      name_prefix      = "http-"
      backend_protocol = "TCP"
      backend_port     = 32080
      port             = 32080
      protocol         = "TCP"
      target_type      = "instance"
      target_id        = var.AUTOSCALING_GROUP_ARN
    }
}
```

```
│ Error: registering targets with target group: ValidationError: Instance ID 'arn:aws:autoscaling:eu-central-1:<account-id>:autoScalingGroup:c4b71856-026e-43f9-bdc1-3b34dfed9a19:autoScalingGroupName/base_arm64-prod-2023020409180057890000000e' is not valid
│       status code: 400, request id: 00a780e5-6fc1-482b-9c78-a20e3859987b
│ 
│   with module.nlb_prod.module.load_balancer.aws_lb_target_group_attachment.this["instance_http"],
│   on .terraform/modules/nlb_prod.load_balancer/main.tf line 507, in resource "aws_lb_target_group_attachment" "this":
│  507: resource "aws_lb_target_group_attachment" "this" {

```

I also checked in the AWS console how this would work, but couldn't get it going for an ASG there either.

**Comment by bryantbiggs** [comment]
ah, I didn't realize it was a separate resource. We'll need to add it https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/autoscaling_attachment

**Comment by rdhar** [comment]
Hi @bryantbiggs, many thanks for the huge contribution towards converting from lists to maps; sure there's some nesting, but the `target_group_key` is great a addition for maintenance and legibility.

In the same vein as @pat-s, I'm looking to attach N-number of targets to a group: specifically, ECS Fargate instances. With module/alb v8, it was possible to simply forward a listener to a target group with `target_type = "ip"`, as [demonstrated in module/ecs v5](https://github.com/terraform-aws-modules/terraform-aws-ecs/blob/master/examples/fargate/main.tf#L203-L218).

Whereas, with module/alb v9, I'm not sure what to pass as the `target_id` for the target group❔ This is further compounded by module/ecs v5 requiring `load_balancer = { service = { target_group_arn = ... } }`, so it needs to be provisioned after the alb. Unless there's a way to pass the capacity provider's ID without creating a circular dependency? 

Thanks for your time.

**Comment by stivi21** [comment]
> The recommendation there for HA would be to attach an autoscaling group

In many places we don't have autoscaling group, but configuration is like that:

```
module "ec2_servers" {
  source  = "terraform-aws-modules/ec2-instance/aws"

  for_each = toset(["0", "1", "2"])
  ...
}
```
and in previous module version I configured it like that:

```
targets = {
        for i, server in module.ec2_servers :
        i => {
          target_id = server.id
          port      = 80
        }
      }
```
Is there any other way to configure it?
Thank you

**Comment by bryantbiggs** [comment]
@rdhar I've updated the ECS module to use the latest ALB module and demonstrate how these two modules integrate https://github.com/terraform-aws-modules/terraform-aws-ecs/pull/130

Next, I'm working on updating the autoscaling module to add the target group attachment resource

**Comment by bryantbiggs** [comment]
traffic source attachment for autoscaling groups is being added here https://github.com/terraform-aws-modules/terraform-aws-autoscaling/pull/248 - please see the updated example in that PR for how to target an autoscaling group from the LB

**Comment by pat-s** [comment]
@bryantbiggs AFAIU the update example is linked to the use of `terraform-aws-autoscaling`. Is there a way to use the `nlb` module with a standalone ASG? In my case, the ASG are created through `aws-terraform-eks` and I can't infer how I can add such an ASG to a `target_group` in this module. 

A quick snippet for this would be highly appreciated, we're still operating on a hardcoded workaround right now.

**Comment by bryantbiggs** [comment]
When using EKS, my first recommendation would be to use the AWS load balancer controller. You can let it create all of the LB resources, or you can create a standalone LB and use the target group attachment only functionality

**Comment by rdhar** [comment]
> @rdhar I've updated the ECS module to use the latest ALB module and demonstrate how these two modules integrate [terraform-aws-modules/terraform-aws-ecs#130](https://github.com/terraform-aws-modules/terraform-aws-ecs/pull/130)

@bryantbiggs Huge thanks once again for sharing those examples, since they're crucial while we're working with Math B (@ROunofF) to optimize our ECS Fargate deployment with TF AWS modules. As we migrate various Node/Python microservices to ECS, attaching each one's target group to the same ALB is the ideal goal.

Now, we've run into a problem with paths. To share more context, there are services **A**, **B**, and **C** that are now routed by `forward`er listener rules to the appropriate target group. But all of these are services route at the "root"-level. In other words, URLs like /health and /docs exist for each service, such as: **A/health**, **B/health**, **C/api/docs** nested paths, etc.

However, the ALB listener rule appears to only be able to forward one level, so it's unable to serve any of target groups other than the default. Here's the specific rule configuration for reference, where I've tried both with and without wildcard `*`.

```hcl
conditions = [{ path_pattern = {
                values = ["/${local.srvA}/*"]
             }}]
actions    = [{
                type = "forward"
                target_group_key = local.srvA
             }]
```

Am I missing something obvious that'd allow me to serve all of service **A**'s endpoints under **example.org/A/health**, then **example.org/C/api/docs**, and so forth?

**Comment by pat-s** [comment]
> When using EKS, my first recommendation would be to use the AWS load balancer controller. You can let it create all of the LB resources, or you can create a standalone LB and use the target group attachment only functionality

Thanks. We'll consider the LB controller for upcoming clusters.
For the current one however, we're now stuck with NLB and the need to reference ASG from the `terraform-aws-eks` module. 

> use the target group attachment only functionality

Could you link to an example here? I've tried searching all docs and examples but I can't figure out how I can dynamically link an ASG to a `target_group` with `type: instance` in NLB v9. If this is not possible at all, this would already be a helpful information.

**Comment by remi-f-artelia** [comment]
Hi there, 
First of all thanks for the conversion from lists to maps. That's a huge progress IMHO and I hope this will be done in many other aws tf modules.
I'm also in the need to have several instance targets in a target group and no possibility to add these instances into a ASG.
So until this is possible I forked your module to proceed as below.

Notes:
- var `target_groups` ==> `target_group_map` in my code
- add a new vars `target_group_instance_attachment_map`  defined as below

```hcl
# Instance | ECS Container
variable "target_group_instance_attachment_map" {
  description = "Map of target group attachment configurations to create"
  type = map(object({
    availability_zone = optional(string, null) # The Availability Zone where the IP address of the target is to be registered. If the private IP address is outside of the VPC scope, this value must be set to all.
    port              = optional(string, null) # The port on which targets receive traffic.
    protocol          = optional(string, null) # The protocol on which targets receive traffic.
    target_arn        = optional(string, null) # The ARN of the target group with which to register targets.
    target_id         = optional(string, null) # The Instance ID, or the container ID of the target. 
  }))
  default = {}
}
```
- add a new vars `target_group_ip_attachment_map`  defined as below

```hcl
# IP Address
variable "target_group_ip_attachment_map" {
  description = "Map of target group attachment configurations to create"
  type = map(object({
    availability_zone = optional(string, null) # The Availability Zone where the IP address of the target is to be registered. If the private IP address is outside of the VPC scope, this value must be set to all.
    port              = optional(string, null) # The port on which targets receive traffic.
    protocol          = optional(string, null) # The protocol on which targets receive traffic.
    target_arn        = optional(string, null) # The ARN of the target group with which to register targets.
    target_id         = optional(string, null) # The IP address of the target. 
  }))
  default = {}
}
```
- add a new vars `target_group_lambda_attachment_map`  defined as below

```hcl
# Lambda
variable "target_group_lambda_attachment_map" {
  description = "Map of target group attachment configurations to create"
  type = map(object({
    availability_zone = optional(string, null) # The Availability Zone where the IP address of the target is to be registered. If the private IP address is outside of the VPC scope, this value must be set to all.
    port              = optional(string, null) # The port on which targets receive traffic.
    protocol          = optional(string, null) # The protocol on which targets receive traffic.
    target_arn        = optional(string, null) # The ARN of the target group with which to register targets.
    target_id         = optional(string, null) # The Lambda function ARN of the target.
  }))
  default = {}
}
```
- add a new vars `target_group_alb_attachment_map`  defined as below

```hcl
# ALB
variable "target_group_alb_attachment_map" {
  description = "Map of target group attachment configurations to create"
  type = map(object({
    availability_zone = optional(string, null) # The Availability Zone where the IP address of the target is to be registered. If the private IP address is outside of the VPC scope, this value must be set to all.
    port              = optional(string, null) # The port on which targets receive traffic.
    protocol          = optional(string, null) # The protocol on which targets receive traffic.
    target_arn        = optional(string, null) # The ARN of the target group with which to register targets.
    target_id         = optional(string, null) # The ALB ARN of the target.
  }))
  default = {}
}
```

Add the following local

```hcl
locals {
target_group_attachment_map = merge(
    { # Instance ID | ECS Container ID
      for k, v in var.target_group_instance_attachment_map : k => merge(
        v,
        {
          target_group_arn = [
            for tg in aws_lb_target_group.this : tg.arn if tostring(tg.port) == v.port && tg.protocol == v.protocol
          ][0]
        }
      )
    },
    { # IP address
      for k, v in var.target_group_ip_attachment_map : k => merge(
        v,
        {
          target_group_arn = [
            for tg in aws_lb_target_group.this : tg.arn if tostring(tg.port) == v.port && tg.protocol == v.protocol
          ][0]
          target_id = join(",", data.dns_a_record_set.this[k].addrs)
        }
      )
    },
    # didn't tested the 2 next options so there are commented in my local
    # { # Lambda ARN
    #   for k, v in var.target_group_lambda_attachment_map : k => merge(
    #     v,
    #     {
    #       target_group_arn = [
    #         for tg in aws_lb_target_group.this : tg.arn if tostring(tg.port) == v.port && tg.protocol == v.protocol
    #       ][0]
    #       target_id = join(",", dns_a_record_set.this.addrs)
    #     }
    #   )
    # },
    # { # ALB ARN
    #   for k, v in var.target_group_alb_attachment_map : k => merge(
    #     v,
    #     {
    #       target_group_arn = [
    #         for tg in aws_lb_target_group.this : tg.arn if tostring(tg.port) == v.port && tg.protocol == v.protocol
    #       ][0]
    #       target_id = join(",", dns_a_record_set.this.addrs)
    #     }
    #   )
    # },
}
```

The resource 'Target Group Attachment' is set as below

```hcl
################################################################################
# Target Group Attachment
################################################################################

resource "aws_lb_target_group_attachment" "this" {
  depends_on = [
    aws_lb_target_group.this,     # I'm not sure this dependancy is really needed since it is set in my local.
    aws_lambda_permission.this
  ]
  for_each = {
    for k, v in local.target_group_attachment_map :
    k => v if local.create && lookup(v, "create_attachment", true)
  }

  # target_group_arn  = aws_lb_target_group.this[each.key].arn
  target_group_arn  = each.value.target_group_arn
  target_id         = each.value.target_id
  port              = try(each.value.target_type, null) == "lambda" ? null : try(each.value.port, var.default_port)
  availability_zone = try(each.value.availability_zone, null)

}
```

I hope this could help to be added in an official feature in order I could use this module and not my fork anymore.

**Comment by bryantbiggs** [comment]
> need to have several instance targets in a target group and no possibility to add these instances into a ASG

I am curious, and want to better understand - why the need for standalone instances like this?

**Comment by remi-f-artelia** [comment]
The solution is provided by an editor which does not allow to work with ASG.
This is not a solution that I approve but I needed this to be available quickly so I didn't have the time to look for a better option for now.

**Comment by bryantbiggs** [comment]
I'm open to hear other use cases but so far all of the feedback and data I've collected has pointed to this being an edge case where "modern practices" are not utilized. Therefore, I think the workaround for these edge cases, if necessary, is to let this module create the target group but without any attachments `create_attachment = false`, and then use the `aws_lb_target_group_attachment` resource to make the necessary attachments. You can do 1:1, use `for_each`, etc., its up to you and your use case. This saves the hassle of all of the iterating deconstructing and re-constructing of mappings to make the attachments (KISS). Something like below:

```hcl
module "alb" {
  source  = "terraform-aws-modules/alb/aws"
  version = "~> 9.0"

  # Truncated ...

  target_groups = {
    example = {
      backend_protocol = "HTTP"
      backend_port     = 80
      target_type      = "instance"

      # Attachments will be made externally
      create_attachment = false
    }
  }
}

resource "aws_lb_target_group_attachment" "external" {
  for_each = {for k,v in aws_instance.example: k => v}

  target_group_arn = module.alb.target_groups["example"].arn
  target_id        = each.value.id
  port             = 80
}

resource "aws_instance" "example" {
  for_each = toset(["1", "2", "3"]) # ¯\_(ツ)_/¯

  ami           = data.aws_ami.ubuntu.id
  instance_type = "t3.micro"
}
```

**Comment by rdhar** [comment]
While it's in the context of ECS Fargate instances instead of EC2, but [here's a working snippet](https://github.com/rdhar/ecs-fargate-demo/blob/main/networks.tf#L78-L83) to demo for-each loop of `target_groups` directly within ALB module, like so:

```hcl
locals {
  services = [
    { name = "alpha"
      port = 80 },
    { name = "bravo"
      port = 80 },
  ]}

target_groups = {
  for service in local.services : service.name => {
    name              = service.name
    backend_port      = service.port
    target_type       = "ip"
    create_attachment = false
}}
```

**Comment by k911** [comment]
@bryantbiggs 

Another valid use case for myself is to create a private ALB that targets VPC endpoint IP addresses, such a scenario is sometimes needed in private networks to handle TLS properly on internal CA certificates and custom private domains, in that case, targets would be bare multiple IP addresses. Currently, there is a big limitation in the module that can be handled by a workaround. 

Another use case is for hybrid cloud environments which also would require IP targets. 

> edge case where "modern practices" are not utilized

Unfortunately in my cases ("IP" target usage) to my current knowledge cannot be solved assuming we still use AWS ELBs so, for me those are not edge cases and I would call them just valid but very rare use cases.

**Comment by jseiser** [comment]
@bryantbiggs 

This module upgrade broke us as well.  We are required per our accreditation to require all traffic to be encrypted, and to have a LB sandwich.  We also require being able to push traffic to 1 of 2 EKS clusters

* 1 Public VPC, with a Public ALB, Fronted by an AWS WAF.
* 2 Private VPC's, each with an EKS cluster, with NGINX ingress creating a private NLB.
* Create VPC Endpoints For each IP/AZ of each Private NLB created by NGINX Ingress
* Weighted Route 443 on the Public ALB to 1 of the 2 Private NLB's depending on which cluster is `active`

```
      targets = {
        nlb1 = {
          port      = 443
          target_id = data.aws_network_interface.nlb0.private_ip
        }
        nlb2 = {
          port      = 443
          target_id = data.aws_network_interface.nlb1.private_ip
        }
        nlb3 = {
          port      = 443
          target_id = data.aws_network_interface.nlb2.private_ip
        }
      }
```

I know this is not an edge case, since its the 3rd place I've been at that handled Blue/Green Clusters this way.

**Comment by bryantbiggs** [comment]
I'm sorry, what exactly is the issue in the example above?

**Comment by jseiser** [comment]
In the previous version of the module, a target group, could have multiple targets defined, in the new version of the module, it does not appear, from what I can tell, that you can attach more than 1 target to a target group.  I cant use your ASG suggestion etc, since our targets are the IP's of the VPC endpoints for the private NLB's.

**Comment by bryantbiggs** [comment]
Ah ok, stating `This module upgrade broke us as well.` is not quite accurate if I understand correctly - its just that the previous behavior is not carried over to the current module. 

In that scenario, if we ignore the number of ways that could break due to the dynamic nature of endpoints and IPs 😅  - does the configuration you provided not work for your use case? Is it merely just trying to avoid multiple target groups? 

There is the alternate route where you can create the attachment as well - similar to the relationship between this module and the ECS module. You would have this module create the target group, but no attachments - and then you can use the attachment resource to attach the appropriate IDs in the fashion of your chosing

**Comment by jseiser** [comment]
Moving to multiple TG's would require a re-accreditation of the system topology.  So we dont have any choice but to re-create the attachments outside of the module.

I was simple demonstrating that, stating `edge case where "modern practices" are not utilized` is not quite accurate

**Comment by an-stark** [comment]
A load balancer that can only attach one instance and not "load balance" between multiple instances does not seem like an edge case in any way.  It is not an LB, it's an SSL termination for one ec2.  You might want to poll the user base before making such changes in the future.

**Comment by bryantbiggs** [comment]
> A load balancer that can only attach one instance and not "load balance" between multiple instances does not seem like an edge case in any way. It is not an LB, it's an SSL termination for one ec2. You might want to poll the user base before making such changes in the future.

1. How do you suggest we poll the user base? The [refactoring PR](https://github.com/terraform-aws-modules/terraform-aws-alb/pull/305) was opened in August and closed in October, that seems like plenty of time for feedback, no? 
2. You can attach multiple instances - you just can't do it direct to multiple instances natively in this module. You will need to either use an ASG and load balance traffic over container of instances, or you can use the attachement resource yourself to attach the long lived instances

**Comment by an-stark** [comment]
1. I guess I will start following these more closely looking for such changes that are not backwards compatible.
2. It seems you traded a "hassle deconstructing mappings" inside the module for "hassle deconstructing mappings" outside the module in my .tf file.

Was just frustrating to run into, I'm sure you can understand.

**Comment by bryantbiggs** [comment]
> changes that are not backwards compatible

That is the definition of a breaking change, FYI. 

> It seems you traded a "hassle deconstructing mappings" inside the module for "hassle deconstructing mappings" outside the module in my .tf file. Was just frustrating to run into, I'm sure you can understand.

Not exactly - and I'll go back to my previous comment. Trying to treat IPs or instances today as pets is not the norm and therefore thats an edge case. We design things for the norm - for the majority of the population. I didn't see any reasoning from yourself as to why you need to map direct to an instance, but those reasons today are dwindling. Even to @jseiser use case (not that his 👎🏽 adds to this conversation) - the load balancer sandwich architecture is a legacy artifact. If you have business requirements that stipulate this, fine, but we cannot satisfy all requests - we aim to satisfy the majority. For those we can't satisfy, we have provided guidance and alternative solutions. Its just frustrating that something that is provided for free - FREE - and people have the audacity to complain and state demands

**Comment by an-stark** [comment]
# For those we can't satisfy, we have provided guidance and alternative solutions.

Your guidance and alternative solution worked, thank you for providing it.

I can appreciate where you are coming from and I may have read intention that was not there, which is hard to gauge in a written medium, I apologize.  You are correct that this is a legacy way of doing things and I will have to figure out a way to make the application move forward and be less pet-like.

**Comment by georambo** [comment]
```
module "alb" {
  source  = "terraform-aws-modules/alb/aws"
  version = "~> 9.0"

  # Truncated ...

  target_groups = {
    example = {
      backend_protocol = "HTTP"
      backend_port     = 80
      target_type      = "instance"

      # Attachments will be made externally
      create_attachment = false
    }
  }
}

resource "aws_lb_target_group_attachment" "external" {
  for_each = {for k,v in aws_instance.example: k => v}

  target_group_arn = module.alb.target_groups["example"].arn
  target_id        = each.value.id
  port             = 80
}

resource "aws_instance" "example" {
  for_each = toset(["1", "2", "3"]) # ¯\_(ツ)_/¯

  ami           = data.aws_ami.ubuntu.id
  instance_type = "t3.micro"
}
```

I guess this solution to have `create_attachment = false`  is not yet implemented?  

i receive the following error if trying to use that

```
ValidationError: A target group ARN must be specified
```

**Comment by an-stark** [comment]
It definitely worked for me, not sure if there is a "listener" section under your Truncated bit, but that is necessary I believe.  I'll grab my config and sanitize it for posting here.

**Comment by an-stark** [comment]
`module "example_nlb" {
  source = "../../modules/lb/"

  name = "example-nlb"

  load_balancer_type    = "network"
  vpc_id                = module.vpc.vpc_id


  listeners = {
    example = {
      port            = 51400
      protocol        = "TLS"
      certificate_arn = module.example_cert.acm_certificate_arn
      forward = {
        target_group_key = "example_zero"
      }
    }
  }

  target_groups = {
    example_zero = {
      name                 = "example-tg"
      protocol             = "TCP"
      port                 = 51400
      target_type          = "instance"
      deregistration_delay = 10

      # Attachments will be made externally
      create_attachment = false
		} 
    }
  }
}

resource "aws_lb_target_group_attachment" "example" {
  for_each = {for k,v in module.ec2_example: k => v}

  target_group_arn = module.example_nlb.target_groups["example_zero"].arn
  target_id        = each.value.id
  port             = 51400
}`

**Comment by georambo** [comment]
I see, yeah, that is what i was doing now, i guess you need to create one target group at least with the module, and additional ones outside the module.  👍

**Comment by bryantbiggs** [comment]
> i guess you need to create one target group at least with the module, and additional ones outside the module.

No - the only requirement from a listener is that there is a default action. I don't know what you are trying to do since not a lot of details are provided, but if you are forwarding to a target group, you can forward to one the module creates or an external target group

**Comment by iT3E** [comment]
> edge case where "modern practices" are not utilized

https://aws.amazon.com/blogs/networking-and-content-delivery/hosting-internal-https-static-websites-with-alb-s3-and-privatelink/  barely a year old pattern presented by AWS themselves which requires multiple IP targets in a single target group.  This would have worked fine in the previous module version, but now the `create_attachment = false` workaround is needed.

**Comment by florianmagnin** [comment]
Hi,
Same for us, this module upgrade broke us as well. We could have more than 1 EC2 to attach to the same TG. Sometimes we cannot use ASG.
Thanks

**Comment by joshsizer** [comment]
IMO, users should have the option to pass in a list of targets for a target group. Here's my use case:

aws-load-balancer-controller in EKS is creating an NLB for Nginx in one account. We need an ALB in another account that forwards traffic to these NLB IPs. The only option for us is to pass a list of NLB IPs to the target group.

I'm fully aware this is far from ideal but it seems to me to be a relatively generic use case for multi-account, multi-EKS cluster setups with the need for a single ingress in a different account (ie, we can't just have the aws-load-balancer-controller create our ALB).

To make things more challenging, we are using Terragrunt, whose base unit is a module. This means we need to create and manage our own custom module that follows the work around that has been proposed. It's not as simple as adding another resource for us, unfortunately

**Comment by bryantbiggs** [comment]
> IMO, users should have the option to pass in a list of targets for a target group

Go ahead, try to figure that one out

**Comment by joshsizer** [comment]
> > IMO, users should have the option to pass in a list of targets for a target group
> 
> Go ahead, try to figure that one out

We can add an optional `targets` key to the `target_groups` variable:

```terraform
target_groups = {
    ex-instance = {
      name_prefix      = "h1"
      protocol         = "HTTP"
      target_type      = "ip"
      port = "80"

      targets          = [{
        target_id        = "10.0.0.0"
      },
      {
        target_id        = "10.0.0.1"
      },
      {
        target_id        = "10.0.0.2"
      }]
    }
  }
```

Add a locals block and modify the `aws_lb_target_group_attachment.this` resource:

```terraform
locals {
  target_group_attachments = flatten([
    for tg_name, tg in var.target_groups : concat(
      # If 'targets' is provided, use it
      [for target in lookup(tg, "targets", []) : {
        tg_name           = tg_name
        target_type       = tg.target_type
        target_id         = target.target_id
        port              = try(target.port, try(tg.port, null))
        availability_zone = try(target.availability_zone, try(tg.availability_zone, null))
      }],
      # If 'targets' is not provided, but 'target_id' is, use 'target_id'
      length(lookup(tg, "targets", [])) == 0 && try(tg.target_id, "") != "" ? [{
        tg_name     = tg_name
        target_type = tg.target_type
        target_id   = tg.target_id
        port        = try(tg.port, null)
        availability_zone = try(tg.availability_zone, null)
      }] : []
    )
  ])
}

resource "aws_lb_target_group_attachment" "this" {
  count = local.create ? length(local.target_group_attachments) : 0

  target_group_arn  = aws_lb_target_group.this[local.target_group_attachments[count.index].tg_name].arn
  target_id         = local.target_group_attachments[count.index].target_id
  port              = try(local.target_group_attachments[count.index].target_type, null) == "lambda" ? null : try(local.target_group_attachments[count.index].port, var.default_port)
  availability_zone = try(local.target_group_attachments[count.index].availability_zone, null)

  depends_on = [aws_lambda_permission.this]
}
```

For a diff, view the MR here: https://github.com/joshsizer/terraform-aws-alb/pull/1/files

I've tested this in my own Terragrunt project, but I'd be happy to make an MR into this repo and do the appropriate testing with the examples directory if this is an acceptable solution.

**Comment by bryantbiggs** [comment]
we are not changing from `for_each` back to `count` - that would be a no go

**Comment by joshsizer** [comment]
> we are not changing from `for_each` back to `count` - that would be a no go

I'd be happy to rework this by changing the `targets` variable to be a map instead of array, like so:

```terraform
target_groups = {
        ex-instance = {
            name_prefix      = "h1"
            protocol         = "HTTP"
            target_type      = "ip"

            targets = {
                "target_1" = {
                    target_id = "10.0.0.0"
                },
                "target_2" = {
                    target_id = "10.0.0.1"
                },
                "target_3" = {
                    target_id = "10.0.0.2"
                }
            }
        }
    }
```

Of course, there would now be a 1:n mapping of target group attachments to target groups, so the keys used in the `for each` would have to combine the target group name and the target name.

Targets don't have names in AWS, so I'm not sure a map would make sense here in general, however if that's your main stipulation  I'd be happy to rewrite this using `for_each`.

**Comment by k911** [comment]
@joshsizer I think toset() could help with having list of targets instead of map and still using foreach

```
for_each = toset(local.array_with_unique_items)
```

**Comment by bryantbiggs** [comment]
> however if that's your main stipulation I'd be happy to rewrite this using `for_each`

I don't have any stipulations per se, but we just did a refactor to move from `count` to `for_each` based on a number of known challenges and issues with the prior approach. So I don't think it make sense to go backwards - plus this would be a breaking change as well

I also want to highlight that the solution thats been proposed will only work if users pass each value statically - you cannot use `toset()` or loop over and unpack results from a data source - keys in any map cannot be dynamic

**Comment by joshsizer** [comment]
> I also want to highlight that the solution thats been proposed will only work if users pass each value statically - you cannot use `toset()` or loop over and unpack results from a data source - keys in any map cannot be dynamic

After some testing I understand the issue that you're getting at. We could do something similar to how each listener references a target group:

```terraform
additional_target_group_attachments = {
  ex-instance-target-1 = {
    target_group = "ex-instance"
    target_id    = "10.0.0.70"
  }
  ex-intance-target-2 = {
    target_group = "ex-instance"
    target_id    = "10.0.0.71"
  }
}
```

```terraform 
resource "aws_lb_target_group_attachment" "additional_target_group_attachments" {
  for_each = { for k, v in var.additional_target_attachments : k => v if local.create }

  target_group_arn  = aws_lb_target_group.this[each.value.target_group].arn
  target_id         = each.value.target_id
  port              = try(each.value.target_type, null) == "lambda" ? null : try(each.value.port, var.default_port)
  availability_zone = try(each.value.availability_zone, null)

  depends_on = [aws_lambda_permission.this]
}
```

This would also be backwards compatible, which is a plus. @bryantbiggs What do you think about this approach?

**Comment by dbaber** [comment]
I don't want to get into the argument of things because I inherited this but what is the "solution" here ( I think they did this because VPCE private IPs were being load balanced and lambdas hung off of that) They even had things pinned to version 7.0.0. No ASGs or anything like that. It does work with 7.0.0, but that seems old.

```
  target_groups = [
    {
      name             = "lx-${local.application}-alb-tg-${local.env}-${local.region_short}"
      backend_protocol = "HTTPS"
      backend_port     = 443
      target_type      = "ip"
      vpc_id = data.aws_vpc.main.id
      health_check = {
        enabled             = true
        interval            = 10
        path                = "/"
        port                = "443"
        healthy_threshold   = 3
        unhealthy_threshold = 3
        timeout             = 5
        protocol            = "HTTPS"
        matcher             = "200,403"
      }
      targets = [
        {
          target_id = data.aws_network_interface.eni_ip_a.private_ip
          port      = 443
        },
        {
          target_id = data.aws_network_interface.eni_ip_b.private_ip
          port      = 443
        },
      ]
    }
  ]
```

**Comment by joshsizer** [comment]
> @joshsizer yes I think we can add support for that (sorry, must have missed the last notification)

Awesome, I’ll work on a PR for this 👍

**Comment by antonbabenko** [comment]
This issue has been resolved in version 9.7.0 :tada:
